core/storage/trie/serialization/codec.hpp

```diff
@@ -6,42 +6,51 @@
 #ifndef KAGOME_TRIE_CODEC_HPP
 #define KAGOME_TRIE_CODEC_HPP
 
-#include "common/blob.hpp"
-#include "common/buffer.hpp"
-#include "storage/trie/node.hpp"
-#include "storage/trie/types.hpp"
+#include "../../../common/blob.hpp"
+#include "../../../common/buffer.hpp"
+#include "../polkadot_trie/trie_node.hpp"
+#include "../types.hpp"
 
 namespace kagome::storage::trie {
   struct TrieNode;
 
   /**
-   * @brief Internal codec for nodes in the Trie. Eth and substrate have
-   * different codecs, but rest of the code should be same.
+   * @brief Internal codec for nodes in the Trie.
    */
   class Codec {
    public:
-    using StoreChildren = std::function<outcome::result<void>(
-        const TrieNode *, common::BufferView, common::Buffer &&)>;
+    using ChildVisitor = std::function<outcome::result<void>(
+        TrieNode const & /* a child node */,
+        common::BufferView /* its merkle value */,
+        common::Buffer && /* the encoded node */)>;
+
+    static constexpr auto NoopChildVisitor =
+        [](auto const &, auto, auto &&) -> outcome::result<void> {
+      return outcome::success();
+    };
 
     virtual ~Codec() = default;
 
     /**
      * @brief Encode node to byte representation and store children
      * @param node node in the trie
-     * @param store_children chidren storer
+     * @param version the version of trie encoding algorithm, preserve as is if
+     * nullopt
+     * @param child_visitor function invoked for every child in a branch node
+     * recursively
      * @return encoded representation of a {@param node}
      */
     virtual outcome::result<common::Buffer> encodeNode(
-        const Node &node,
-        StateVersion version,
-        const StoreChildren &store_children) const = 0;
+        const TrieNode &node,
+        std::optional<StateVersion> version,
```

Comment:
`StateVersion::V0` should be used explicitly
```suggestion
        StateVersion version,
```